streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,283 @@
+        raise ValueError(f'{spark_data_type} is not supported by MDSwriter')
+
+    if user_defined_cols is not None:  # user has provided schema, we just check if mds supports the dtype
+        mds_supported_dtypes = set(mapping_spark_to_mds.values())
+        for col_name, user_dtype in user_defined_cols.items():
+            if col_name not in dataframe.columns:
+                raise ValueError(
+                    f'{col_name} is not a column of input dataframe: {dataframe.columns}')
+            if user_dtype not in mds_supported_dtypes:
+                raise ValueError(f'{user_dtype} is not supported by MDSwriter')
+
+            actual_spark_dtype = dataframe.schema[col_name].dataType
+            mapped_mds_dtype = map_spark_dtype(actual_spark_dtype)
+            if user_dtype != mapped_mds_dtype:
+                raise ValueError(f"Mismatched types: {col_name} is {mapped_mds_dtype} in DataFrame but {user_dtype} in user_defined_cols")
+        return None
+
+    schema = dataframe.schema
+    schema_dict = {}
+
+    for field in schema:
+        dtype = map_spark_dtype(field.dataType)
+        if dtype in _encodings:
+            schema_dict[field.name] = dtype
+        else:
+            raise ValueError(f'{mds_dtype} is not supported by MDSwriter')
+    return schema_dict
+
+
+def do_merge_index(partitions: Iterable,
+                   mds_path: Union[str, Tuple[str, str]]):
+    """Merge index.json from partitions into one for streaming.
+
+    Args:
+        partitions (Iterable): partitions that contain pd.DataFrame
+        mds_path (Union[str, Tuple[str, str]]): (str,str)=(local,remote), str = local or remote based on parse_uri(url) result
+    Return:
+        None
+    """
+    if not partitions:
+        logger.warning("No partitions exist, no mergei index")
+        return
+
+    shards = []
+
+    for row in partitions:
+        mds_partition_index = f'{row.mds_path}/{get_index_basename()}'
+        mds_partition_basename = os.path.basename(row.mds_path)
+        obj = json.load(open(mds_partition_index))
+        for i in range(len(obj['shards'])):
+            shard = obj['shards'][i]
+            for key in ['raw_data', 'zip_data']:
+                if shard.get(key):
+                    basename = shard[key]['basename']
+                    obj['shards'][i][key]['basename'] = os.path.join(mds_partition_basename,
+                                                                     basename)
+        shards += obj['shards']
+
+    obj = {
+        'version': 2,
+        'shards': shards,
+    }
+
+    if isinstance(mds_path, str):
+        mds_index = os.path.join(mds_path, get_index_basename())
+    else:
+        mds_index = os.path.join(mds_path[0], get_index_basename())
+
+    with open(mds_index, 'w') as out:
+        json.dump(obj, out)
+
+
+def dataframeToMDS(dataframe: DataFrame,
+                   merge_index: bool = True,
+                   mds_kwargs: Optional[Dict[str, Any]] = None,
```

Comment:
Instead of nesting kw_args, why not just do `**mds_kwargs` at the end, and then any args that are passed in the end are routed to MDSWriter? As long as no kwargs conflict with dataframeToMDS it should be fine.

Reply:
but mds_kwargs[out] is different for each partition. @Skylion007